### PR TITLE
Manage version configurations client versions

### PR DESCRIPTION
**Pull request by csduarte**
#### Summary
Added config values for client versions and return client versions in ping response. These values will be used by app clients to display suggest the user to update or force the user to update:

    "ClientRequirements": {
        "AndroidLatestVersion": "",
        "AndroidMinVersion": "",
        "DesktopLatestVersion": "",
        "DesktopMinVersion": "",
        "IosLatestVersion": "",
        "IosMinVersion": ""
    },

Clients will hit the ping endpoint: `curl -i http://localhost:8065/api/v4/system/ping` and based on the the min version or latest version it would ask for the update:

screen shot 2017-08-15 at 4 02 09 pm

Added ability to manage client version through System Console. Also, added client versions to diagnostics.

screen shot 2017-08-15 at 3 59 13 pm

#### Checklist
- [ ] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (required for all new APIs)
- [ ] All new/modified APIs include changes to the drivers
- [ ] Has enterprise changes (please link)
- [ ] Has UI changes
- [x] Includes text changes and localization file
- [ ] Touches critical sections of the codebase (auth, upgrade, etc.)

via @dmeza

### CODE REVIEW

**Comment by enahum** [comment]
Before reviewing what happens if those settings aren't set? What should be the behavior in the app? What about previous server versions that do not have this settings?

What if the user does ugrade the app but the server has not updated the latests version?

**Comment by dmeza** [comment]
@enahum 
- If the settings aren't set then do nothing.
- For server versions that don't have this setting it would be treated like empty.
- If the user has a newer version than the server, then do nothing.
- The versions are integers separated by dots (it controls if not an integer) and it compares the matching indexes.

This is the logic used for update messages on Android client:
`
private void checkVersion(String minVersion, String latestVersion) {
        if(compareVersionNames(BuildConfig.VERSION_NAME, latestVersion) < 0 && compareVersionNames(BuildConfig.VERSION_NAME, minVersion) >= 0) {
           showUpdateAvailableMessage(latestVersion);
        } else if (compareVersionNames(BuildConfig.VERSION_NAME, minVersion) < 0) {
            showMandatoryUpdateMessage(latestVersion);
        } else {
            runInitialLoad();
        }
    }
`

`
/**
     * Compares two version names
     * @param lVersionName local version name
     * @param rVersionName remote version name
     * @return 0 if both are equal,
     * -1 if lVersionName < rVersionName
     * 1 if lVersionName > rVersionName
     */
    private int compareVersionNames(String lVersionName, String rVersionName) {
        int res = 0;
        // Check if remote Version is invalid do nothing
        if (rVersionName == null || rVersionName.isEmpty()) {
            return 0;
        }

        String[] lVersionTokens = lVersionName.split("\\.");
        String[] rVersionTokens = rVersionName.split("\\.");

        int maxAvailableIndex = Math.min(lVersionTokens.length, rVersionTokens.length);

        for (int i = 0; i < maxAvailableIndex; i ++) {
            // If local version has a non integer character, require update
            if (!lVersionTokens[i].matches("\\d+")) {
                return -1;
            }
            // If remote version has a non integer character do not show an error
            if (!rVersionTokens[i].matches("\\d+")) {
                return 0;
            }
            int lVersionPart = Integer.valueOf(lVersionTokens[i]);
            int rVersionPart = Integer.valueOf(rVersionTokens[i]);

            if (lVersionPart < rVersionPart) {
                res = -1;
                break;
            } else if (lVersionPart > rVersionPart) {
                res = 1;
                break;
            }
        }
        // if local version is shorter than remote version at this point and the previous component
        // were equal then require update.
        if (res == 0 && lVersionTokens.length < rVersionTokens.length) {
            res = -1;
        }
        return res;
    }
`

File: webapp/components/admin_console/client_versions_settings.jsx [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+            iosLatestVersion: config.ClientRequirements.IosLatestVersion,
+            iosMinVersion: config.ClientRequirements.IosMinVersion
+        };
+    }
+
+    renderTitle() {
+        return (
+            <h3>
+                <FormattedMessage
+                    id='admin.security.client_versions'
```
Comment (enahum) [inline webapp/components/admin_console/client_versions_settings.jsx]: all this strings have to be present in `webapp/i18n/en.json`

**Comment by jasonblais** [comment]
@dmeza Is the admin required to update the "LatestVersion" numbers? I assume they don't update dynamically as new mobile/desktop app versions are released.

Are the iOS and Android versions for the new react native apps or the Classic apps?

For the desktop app, what would happen if I'm running two servers where
 1) Running MM server v3.6.

**Comment by dmeza** [comment]
@jasonblais 
1 - yes the admin required to update Version numbers. It's not automatic/dynamic as client versions are released.
2 - All the MM clients would need to be updated to be notified (required/suggested) to update (React native clients, native clients or desktop).
3 - For desktop it would require you to update to the highest min version set by any of the servers you connect to. In your example it would be version 3.6 set on TAB 2:
a) TAB 1: Running MM server v3.6. Latest desktop version is 3.5; minimum is 3.3
b) TAB 2: Running MM server v4.1. Latest desktop version is 3.7; minimum is 3.6.
Let me know if that makes sense.

**Comment by jasonblais** [comment]
Thanks for the quick reply,

1 - Would it be possible for us to set the "latest" version numbers in each release on the Mattermost side (say as part of a release process), which dynamically updates the config setting?

If we did that, maybe the config setting wouldn't even be needed?

Let me know if that makes sense or not.

2 - I'm not sure what you mean..

So for mobile apps (both iOS and Android), we have to different apps.
 - "Mattermost", built with React Native: https://itunes.apple.com/us/app/mattermost/id1257222717?mt=8
 - "Mattermost Classic", which is just a web wrapper we used to have: https://itunes.apple.com/us/app/mattermost-classic/id984966508?mt=8

Which app do the version numbers apply to? Since the version numbers on those apps are different ("Mattermost" is on v1.X; "Mattermost Classic" is on v3.X)

3 - Yeah that makes sense. I think there could be issues with compatibility (say, desktop version 4 may not be compatible with server version 3). But that's a different topic not really related to this.

4 - What does the "latest version" setting do? I understand the minimum version forces you to upgrade. Is the latest version just used to say "please upgrade to vX.Y"

5 - What happens if "latest version" is < "minimum version"

6 - What happens if "minimum version" > latest release (e.g. someone puts desktop app v10 --> could an admin effectively use that to prevent someone from using the app?)

I had a couple of other questions, but they slipped my mind. I might ask them later.

Thanks for patiently answering my questions :) This is such a great feature to have, thanks for contributing this

**Comment by dmeza** [comment]
@jasonblais the idea for this feature is to advise (latest version) or require (minimum version) the client user to update. Maybe it should be an enterprise feature because it's thought for customers that need to advance their apps independently so their latest and min versions could diverge from MM. cc @csduarte @jarredwitt

1 - Based on the previous explanation I think it would be needed, but there could be a latest from MM.

2 - I get what you mean now. Maybe we need different versions for `react-native` and `classic` ios and android. Is classic going to be deprecated later on? That was my assumption and that's why I didn't take it into account. Let me know if I should add the different config values.

4 - Yes, `latest version` is thought to let users know there's a new version and recommends the update, but lets the user continue without updating.

5  and 6 - would depend on the specific code changes that are implemented to each native client. I think it should be standardized and it's open for discussion. In the example implementation i pasted in the comment:
https://github.com/mattermost/platform/pull/7220#issuecomment-322666928

The comparison of the native client version and the server latest version or min version would prime. It wouldn't matter that "latest version" is < "minimum version" or "minimum version" > latest release.  If native client version < "minimum version" and "minimum version" is actually not an available version then the admin is effectively using that to prevent all users from using that native app.

Please keep the questions coming. All these changes imply changes to native apps, so I think it's important to be clear on the value provided.

**Comment by dmeza** [comment]
@enahum added the translation strings.

**Comment by jasonblais** [comment]
Hey @dmeza, 

I also shared a couple of questions via DM to pre-release.mattermost.com, but I'll also respond to the above here:

1 - Your description makes complete sense. Thank you for it.

For the settings, wondering if something like the following would be an option (I'm brainstorming options, not saying we should go this route)

 - Config setting for minimums, as you've implemented
 - Config setting for latest version, as you've implemented
    - If this config setting is blank, use a hard-coded value set by Mattermost, e.g. "3.7". This version number can then be updated whenever a new version of the apps is released
    - If this config setting is NOT blank, e.g. "3.5", then the hard-coded value is overwritten and the user is recommended to download v3.5. This persists until the admin manually changes the config setting, or if they set the field blank (in which case the user is recommended to download the latest app version)

2 - `react-native` app is preferred. We don't need a setting for the `classic` app as they're in maintenance mode. Good to know this is how it's implemented.

3 - 

4 - Excellent.

5 & 6 - I think those make sense. I'll test those cases out also to see.

7 - I tried testing this PR but couldn't get it to work. In the attached, min version is set to v3.8 but my current desktop app version is 3.7.

When do the update notices get triggered to the user? When they install or re-open the app?

image

**Comment by dmeza** [comment]
Hey @jasonblais,
1) Yes, that can be done. Where would the hard-coded value be? I wouldn't recommend it as hardcoded. Since these are client versions for desktop, IOs, Android, React native would be updated independently of platform they should be in some sort of a remote service or separate config file.

7) All the client apps need the updates to be able to check versions. That hasn't happened yet.

Let me know if you have any other questions.

**Comment by jasonblais** [comment]
Hey @csduarte @dmeza 

We discussed internally, and we're okay merging this PR as-is. There will be some work involved to finish the feature (e.g. build the connection to the apps), but most of the server work is here. We'll just call the feature "Experimental" until then.

As the final piece before dev review, can you help hide the System Console UI settings? You can keep the code, just have it "hidden" so it doesn't show up in the UI. Let me know if you have questions about it.

**Comment by dmeza** [comment]
Hey @jasonblais, `Client Versions` link is now hidden in the System Console, but accessible via url: `/admin_console/security/client_versions`

screen shot 2017-08-24 at 4 17 14 pm

**Comment by jasonblais** [comment]
Thank you @dmeza

**Pull request merged**